File: problems/1143.0_Longest_Common_Subsequence.py
'''
DP
T: O(MN)
S: O(MN)

Runtime: 384 ms, faster than 87.68% of Python3 online submissions for Longest Common Subsequence.
Memory Usage: 21.9 MB, less than 80.84% of Python3 online submissions for Longest Common Subsequence.
'''
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n1, n2 = len(text1), len(text2)
        dp = [[0] * (n1 + 1) for _ in range(n2 + 1)]
        for i in range(n2):
            for j in range(n1):
                if text1[j] == text2[i]:
                    dp[i + 1][j + 1] = dp[i][j] + 1
                else:
                    dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j])
        
        return dp[-1][-1]

'''
DP

Runtime: 419 ms, faster than 90.80% of Python3 online submissions for Longest Common Subsequence.
Memory Usage: 22 MB, less than 71.29% of Python3 online submissions for Longest Common Subsequence.
'''
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m, n = len(text1), len(text2)
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(m):
            for j in range(n):
                if text1[i] == text2[j]:
                    dp[i + 1][j + 1] = dp[i][j] + 1
                else:
                    dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j])
        return dp[-1][-1]

File: java/1143.0_Longest_Common_Subsequence.java
/*
DP, T: O(MN), S: O(MN)

执行用时：14 ms, 在所有 Java 提交中击败了13.50% 的用户
内存消耗：45.2 MB, 在所有 Java 提交中击败了42.38% 的用户
通过测试用例：45 / 45
*/

class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int m = text1.length(), n = text2.length();
        int[][] f = new int[m + 1][n + 1];
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                if (text1.charAt(i - 1) == text2.charAt(j - 1)) {
                    f[i][j] = f[i - 1][j - 1] + 1;
                }
                f[i][j] = Math.max(f[i - 1][j], f[i][j]);
                f[i][j] = Math.max(f[i][j - 1], f[i][j]);
            }
        }
        return f[m][n];
    }
}
